### PR TITLE
BUG fix for (sum/avg/stddev)_over_time & zscore promql functions

### PR DESCRIPTION
**Pull request by nikitag55**
This commit addresses below mentioned issues:

1.) sum_over_time()

If the sequence contains all NaNs, current implementation is giving 0 as result. Instead, it should give NaN as result.

2.) avg_over_time()

If the sequence contains all NaNs, current implementation is giving 0 as result. Instead, it should give NaN as result.

3.) stddev_over_time()

On debugging above functions 1 & 2:

sum_over_time() function will calculate the sum of only non-NaN values and ignore all the NaN values which is CORRECT.

Similarly, avg_over_time() function is also computing the average as  sum of non-NaN values / count of non-NaN values in the sequence which is CORRECT

For stddev_over_time() computation, we need sum, squared sum, avg and count of the non-NaN values in the sequence.

Current code CORRECTLY calculates the  sum, squared sum by making use of non-NaN values.

But it has following two issues:

1.) count in this function is counting ALL the values irrespective of whether they are NaN or not NaN -> this will cause a problem when sequence is a combination of both NaN and not NaN values.

2.) while loop is used to iterate in the sequence, but the variable used to traverse the entire sequence which does work like i+=1, is done ONLY when value is not NaN -> this will cause a problem when sequence is a combination of both NaN and not NaN values because in order to traverse the ENTIRE sequence, we have to this  i+=1 always, irrespective of whether it is NaN or not NaN.

4.) z_score():

Because z_score() is based on avg_over_time() & stdev_over_time(), it is also giving WRONG results when sequence contains both NaN and not NaN values or only NaNs.

It works FINE when there are all not NaNs in the sequence.

Other improvements :

1.)  For unit testing, added more and more test-data sequences for sum/avg/stddev_over_time() & zscore which are combination of NaN & not NaN’s, empty sequence, or only NaN’s sequence.

2.) For unit testing, added zscore simple math function in the test file to see if the result given by promql zscore is correct or not.

3.) For unit testing, there are 3 functions in the test file : avg, squared sums & stdvar.
Limitation of these functions is : If the sequence contains combination of NaN and not NaN values, the computed result will be NaN only.
So, these functions are not good for the sequences which contains NaNs in it.

So in the test file, I have added 4 other functions for computing sum, squared sums, avg, stdvar, count ; which works well for the test-cases when sequences have NaNs wherein NaN is just ignored to compute any sort of result and have used the same to test sum_over_time(), avg_over_time(), stddev_over_time() with variety of test-cases.

4.) Earlier, for ZScoreChunkedFunction implementation an abstract class was also created. Removed it and made use of existing VarOverTimeChunkedFunctionD to achieve the result.

5.) Earlier, for Zscore implementation, LastSampleChunkedFunctionWithNanD was also added in RangeFunctions. Removed it and made use of existing VarOverTimeChunkedFunctionD to achieve the result.

**Pull Request checklist**

- [ ] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

**New behavior :**

**BREAKING CHANGES**

If this PR contains a breaking change, please describe the impact and migration
path for existing applications.
If not please remove this section.

Breaking changes may include:
- Any schema changes to any Cassandra tables
- The serialized format for Dataset and Column (see .toString methods)
- Over the wire formats for Akka messages / case classes
- Changes to the HTTP public API
- Changes to query parsing / PromQL parsing

**Other information**:

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 1]
```diff
@@ -155,11 +155,16 @@ class SumOverTimeChunkedFunctionD extends SumOverTimeChunkedFunction() with Chun
                                 doubleReader: bv.DoubleVectorDataReader,
                                 startRowNum: Int,
                                 endRowNum: Int): Unit = {
-    // NaN values are ignored by default in the sum method
-    if (sum.isNaN) {
-      sum = 0d
+    val it = doubleReader.iterate(doubleVectAcc, doubleVect, startRowNum)
```
Comment (TanviBhavsar) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Removing doubleReader.sum and iterating will make it slower. Sum function is optimized based on type of data like for DeltaDeltaConstDataReader, DeltaDeltaIterator.
You can do something like:
```
val sumVector = doubleReader.sum(doubleVectAcc, doubleVect, startRowNum, endRowNum)
 if (!sumVector.isNaN && sum.isNaN)
   sum = 0d
 sum += sumVector
```

Comment (nikitag55) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Correct. 

But above snippet will work correctly only if `sum` value inside `doubleReader.sum` is initialized with `Double.NaN` as `sum` both `doubleReader.sum` and `SumOverTimeChunkedFunctionD` is initialized with 0. Not sure, if it is set to 0d for some reason or not on both the places.

So, for PR thought to at least present the idea/bug with the changes done. I also think to make all the required changes in `doubleReader.sum` only.

Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Yes as discussed initial sum should be NaN in all implementations of DoubleVectorDataReader.sum. We cannot make all changes in doubleReader.sum as we need to store sum of all chunks in SumOverTimeChunkedFunction. A window can have multiple chunks, so addTimeLongChunks can be called multiple times. We need to update SumOverTimeChunkedFunction sum for every chunk

Comment (velvia) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: As suggested, please move this fix into DoubleReader.sum itself -- the variant that reads from double vectors. (`DoubleVectorDataReader`)  This way if the vector is integral, we use the faster sum which doesn't need to test for NaN.

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala [hunk 2]
```diff
@@ -146,20 +222,39 @@ class AggrOverTimeFunctionsSpec extends RawDataWindowingSpec {
   val numIterations = 10
 
   it("should correctly aggregate sum_over_time using both chunked and sliding iterators") {
-    val data = (1 to 240).map(_.toDouble)
```
Comment (velvia) [inline query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: I see that new data part was added to existing test case.  Can we please separate out the new data to a new test case?  Something about checking for NaN's.

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala [hunk 3]
```diff
@@ -234,64 +329,101 @@ class AggrOverTimeFunctionsSpec extends RawDataWindowingSpec {
   }
 
   it("should aggregate count_over_time and avg_over_time using both chunked and sliding iterators") {
-    val data = (1 to 500).map(_.toDouble)
+    val test_data1 = Seq(
```
Comment (velvia) [inline query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: Here too

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 4]
```diff
@@ -408,15 +407,15 @@ class AvgOverTimeFunction(var sum: Double = Double.NaN, var count: Int = 0) exte
...
   override final def reset(): Unit = {
-    sum = Double.NaN;
-    count = 0d
+    sum = Double.NaN
+    count = 0
   }
 
   final def apply(endTimestamp: Long, sampleToEmit: TransientRow): Unit = {
-    sampleToEmit.setValues(endTimestamp, if (count > 0) sum/count else if (sum.isNaN()) sum else 0d)
+    sampleToEmit.setValues(endTimestamp, sum / count)
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: I think it's still safer to have the divide by zero check here.  I see that in reset() the count is still 0, so there is still this possibility.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 5]
```diff
@@ -493,44 +492,44 @@ class StdVarOverTimeFunction(var sum: Double = 0d,
...
-    count += (endRowNum - startRowNum + 1)
-    sum += _sum
-    squaredSum += _sqSum
   }
 }
 
 class StdDevOverTimeChunkedFunctionD extends VarOverTimeChunkedFunctionD() {
   final def apply(endTimestamp: Long, sampleToEmit: TransientRow): Unit = {
-    val avg = if (count > 0) sum/count else 0d
+    val avg = sum/count
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Here too, I think retaining divide by 0 is good idea

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala [hunk 6]
```diff
@@ -173,8 +263,8 @@ class AggrOverTimeFunctionsSpec extends RawDataWindowingSpec {
       val chunkedIt = chunkedWindowItHist(data, rv, new SumOverTimeChunkedFunctionH(), windowSize, step)
       chunkedIt.zip(data.sliding(windowSize, step).map(_.drop(1))).foreach { case (aggRow, rawDataWindow) =>
         val aggHist = aggRow.getHistogram(1)
-        val sumRawHist = rawDataWindow.map(_(3).asInstanceOf[bv.MutableHistogram])
-                                      .foldLeft(emptyAggHist) { case (agg, h) => agg.add(h); agg }
+        val sumRawHist = rawDataWindow.map(_ (3).asInstanceOf[bv.MutableHistogram])
```
Comment (velvia) [inline query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: space?

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala (hunk 4 above)
Comment (nikitag55) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: yes, but Double.NaN/0 will always be double.NaN. Also, it's not possible for sum to be some value and count to be 0.

Comment (nikitag55) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: also, when count in reset() is 0, sum is Double.NaN. So, Double.NaN/0 = Double.NaN.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala (hunk 5 above)
Comment (nikitag55) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Same reason as above.

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala (hunk 6 above)
Comment (nikitag55) [reply query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: oops. Fixed it.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala (hunk 4 above)
Comment (nikitag55) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Still you want the change?

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala [hunk 7]
```diff
@@ -388,21 +479,21 @@ class AggrOverTimeFunctionsSpec extends RawDataWindowingSpec {
 
       data.sliding(windowSize, step).map(_.length - 2).toBuffer.zip(aggregated2).foreach {
         case (val1, val2) => if (val1 == -1) {
-                                val2.isNaN shouldEqual (true) // window does not have any element so changes will be NaN
-                              } else {
-                                val1 shouldEqual (val2)
-                              }
+          val2.isNaN shouldEqual (true) // window does not have any element so changes will be NaN
```
Comment (TanviBhavsar) [inline query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: Restore indentation

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala [hunk 8]
```diff
@@ -516,32 +607,65 @@ class AggrOverTimeFunctionsSpec extends RawDataWindowingSpec {
...
 
-    var rv3 = timeValueRV(data3)
-    val chunkedIt3 = new ChunkedWindowIteratorD(rv3, 160000, 100000, 180000, 100000,
-      new ZScoreChunkedFunctionD(), queryConfig)
-    val aggregated3 = chunkedIt3.map(_.getDouble(1)).toBuffer
-    aggregated3(0).isNaN shouldBe true
+    (0 until numIterations).foreach { x =>
+      val windowSize = rand.nextInt(100) + 10
+      val step = rand.nextInt(50) + 5
+      info(s"  iteration $x  windowSize=$windowSize step=$step")
```
Comment (TanviBhavsar) [inline query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: Extra space

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala [hunk 9]
```diff
@@ -194,11 +284,11 @@ class AggrOverTimeFunctionsSpec extends RawDataWindowingSpec {
       chunkedIt.zip(data.sliding(windowSize, step).map(_.drop(1))).foreach { case (aggRow, rawDataWindow) =>
         val aggHist = aggRow.getHistogram(1)
         val sumRawHist = rawDataWindow.map(_(4).asInstanceOf[bv.MutableHistogram])
-                                      .foldLeft(emptyAggHist) { case (agg, h) => agg.add(h); agg }
+          .foldLeft(emptyAggHist) { case (agg, h) => agg.add(h); agg }
         aggHist shouldEqual sumRawHist
 
         val maxMax = rawDataWindow.map(_(3).asInstanceOf[Double])
-                                  .foldLeft(0.0) { case (agg, m) => Math.max(agg, m) }
+          .foldLeft(0.0) { case (agg, m) => Math.max(agg, m) }
```
Comment (TanviBhavsar) [inline query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: Restore indentation

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala (hunk 7 above)
Comment (nikitag55) [reply query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: done

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala (hunk 8 above)
Comment (nikitag55) [reply query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: Actually this has come from develop code : https://github.com/filodb/FiloDB/blob/606a2f301a9265d893e76ac031ec506fe3f8f75a/query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala#L378

I have fixed everywhere though.

File: query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala (hunk 9 above)
Comment (nikitag55) [reply query/src/test/scala/filodb/query/exec/rangefn/AggrOverTimeFunctionsSpec.scala]: done

**Comment by TanviBhavsar** [comment]
@nikitag55 You have changed changed sum += doubleReader.sum to sum = doubleReader.sum
A window can have multiple chunks. So sum, count, avg and stdDev would only have the last chunk result.  Refer my previous comment for details.

**Comment by nikitag55** [comment]
> @nikitag55 You have changed changed sum += doubleReader.sum to sum = doubleReader.sum
> A window can have multiple chunks. So sum, count, avg and stdDev would only have the last chunk result. Refer my previous comment for details.

@TanviBhavsar Earlier I couldn't make the test cases such that window comprises of multiple chunks. I figured it out and observed how `addTimeDoubleChunks` will be called several times when there are several chunks in the window. Made changes as required to address such behavior. 


Please review.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala (hunk 4 above)
Comment (vishramachandran) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: We have a TODO to represent no-value using a special NaN instead of regular NaN. We have to check if operations on that NaN result in the same NaN. Given this, it may be better to retain the explicit check.

Comment (nikitag55) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Added the check @vishramachandran.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala (hunk 5 above)
Comment (nikitag55) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Added the check @velvia . Another issue I found is that in stddev, stdvar and zscore, that the count check is done just for computing avg and not for computing stddev & stdvar which is in the newline above the line you pointed here. The fix I added takes care of all now. Thanks for pointing out.

**Comment by nikitag55** [comment]
Addressed all the changes. Please review @velvia @vishramachandran @TanviBhavsar

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala (hunk 4 above)
Comment (velvia) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: We can't rely on NaN.  It is possible, if it is an Integer encoded column, to end up with a sum of 0 and count of 0.    Anyways will have a look.
